### ISSUE TITLE
(performance) Iterating over hash loop join should be done in parallel with `futures::join`

### ISSUE THREAD

**Issue by jon-chuang** [singularity-data/risingwave#2428]
**Is your feature request related to a problem? Please describe.**
Currently, we are doing the following:

lookup cache -> await lookup s3 -> lookup cache -> await lookup s3

**Describe the solution you'd like**

We should do the following: 

lookup cache -> await lookup s3 
                          lookup cache -> await lookup s3
                                                     lookup cache -> await lookup s3
                                                     
To better take advantage of async.

Issues:
1. currently, the outer body probably takes hold of a mutable references. It may be challenging to share such resources amongst loop bodies concurrently. The solution may be mutexes but trying not to hold mutex locks across await boundaries.

**Describe alternatives you've considered**
NIL

**Additional context**
Related: https://github.com/singularity-data/risingwave/issues/1870

**Comment by BugenZhao** [comment]
Good catch! FYI, we've previously also applied this idea in #568.

**Comment by jon-chuang** [comment]
Yes, I agree with the approach of fetching all unique keys in advanced first. This can help us to avoid complicated concurrent async logic in the loop body.

**Comment by skyzh** [comment]
+1 for this! One thing special about join is that it process one chunk at a time, while HashAgg process one barrier at a time. We can first improve `next` function to fetch all join states for a chunk, and later optimize it to be applied to all chunks in one barrier.

https://github.com/singularity-data/risingwave/pull/568 should also be a good example. As everything is done in a single thread, we can use Option::take to transfer ownership. Therefore, we won't use dashmap. I previously found several soundness issues in dashmap, I'm not sure if it is fully fixed for now.

**Comment by yuhao-su** [comment]
Another difference between agg and join is 1 group key in agg produce corresponds to 1 row while 1 hash key corresponds to multiple rows. Anyway, would you like to give it a try? You can assign this issue to yourself if so.

**Comment by jon-chuang** [comment]
Ok, but I'll wait for https://github.com/singularity-data/risingwave/pull/2357 to be merged first.

**Comment by jon-chuang** [comment]
Cache locality is probably better as well if we have no awaits in the loop body.
